crates/wasi-http/src/http_impl.rs

```diff
@@ -92,104 +58,66 @@ impl WasiHttp {
             crate::types::Method::Patch => Method::PATCH,
             _ => bail!("unknown method!"),
         };
+        let mut uri = Uri::builder()
+            .authority(request.authority.as_str())
+            // NOTE: this is broken, but will be fixed by `wasi-http` dependency update
+            .path_and_query(request.path.to_owned() + &request.query);
+        match &request.scheme {
+            Some(Scheme::Http) => uri = uri.scheme("http"),
+            Some(Scheme::Https) => uri = uri.scheme("https"),
+            Some(scheme) => bail!("unsupported scheme `{scheme:?}`"),
+            _ => {}
+        }
+        // NOTE: This does not belong here, the complete struct should have been constructed
+        // on request creation
+        let uri = uri.build().context("failed to build URI")?;
 
-        let scheme = match request.scheme.as_ref().unwrap_or(&Scheme::Https) {
-            Scheme::Http => "http://",
-            Scheme::Https => "https://",
-            // TODO: this is wrong, fix this.
-            _ => panic!("Unsupported scheme!"),
-        };
+        let mut req = Request::builder()
+            .method(method)
+            .uri(uri)
+            .header(hyper::header::HOST, &request.authority);
```

Comment:
Should we really be injecting headers on behalf of the caller?

Reply:
I agree, the responsibility of setting the Host header should be on the user of this interface, rather than on the implementer.

Reply:
I agree as well, this logic was simply moved from existing implementation, will update